### ISSUE TITLE
[Firefox] Duplicate cookies in v2.12.4

### ISSUE THREAD

**Issue by bumpySnake** [violentmonkey/violentmonkey#771]
### What is the problem?
A recent change (probably in commit 335bf9660799a298b8776e98fca1e04533d8204a) causes GM_xmlhttpRequest to send a duplicate set of cookies, separated from the first set by a comma.

In private browsing mode, GM_xmlhttpRequest will send cookies which originated from both a private, and non-private context if either:
1. A cookie has been received in both private and non-private browsing mode
2. A cookie has been received in private browsing mode, and The GM_xmlhttpRequest itself receives a cookie while in private browsing mode

Usage of non-private cookies in private mode existed prior to v2.12.4. Usage of current private session cookies did not exist.

### How to reproduce it?
1. Serve a response with a cookie from any toy server
2. Make a bare-minimum GM_xmlhttpRequest to that server
3. Observe the raw headers from either client or server side

Example server path using Flask:
```
@app.route('/<cookiename>/')
def custom_cookie(cookiename):
  print(request.headers.get('Cookie'))
  response = Response('')
  response.set_cookie(cookiename, b'0')
  return response
```
Example GM_xmlhttpRequest:
`GM_xmlhttpRequest({method:'GET', url: 'http://localhost:8080/fromaddon/'});`

Cookies visiting `localhost:8080/frombrowser/` twice in normal mode:
```
None //first browser visit
frombrowser=0, frombrowser=0; //first addon visit
frombrowser=0; fromaddon=0; //second browser visit
frombrowser=0; fromaddon=0, frombrowser=0; fromaddon=0; //second addon visit
```
Cookies visiting `localhost:8080/frombrowser/` twice in private mode:
```
None //first browser visit
frombrowser=0; //first addon visit
frombrowser=0; //second browser visit
fromaddon=0, frombrowser=0; //second addon visit
```

### What is the expected result?
Request should be sent with a single set of the appropriate cookies for the page.

`keya=valuea; keyb=valueb;`

Cookies should also be sent from either a private or non-private context in private browsing mode. Not both.

### What is the actual result?
Cookies are duplicated.

`keya=valuea; keyb=valueb, keya=valuea; keyb=valueb;`

Cookies from both private and non-private contexts are sent in private browsing mode.

### Environment
- Browser: Firefox
- Browser version: 71.0
- Violentmonkey version: 2.12.4
- OS: Windows 10

**Comment by tophf** [comment]
Can you make a simpler repro using the barebone SimpleHTTPServer or similar for both cases? Neither me nor gera2ld use Firefox so it'd be very helpful.

Currently my guess is that for some reason Firefox sends a `cookie` header in your use case and Violentmonkey adds another one, because I can't imagine how a comma-joined list would appear otherwise. Can you inspect the request in some MitM proxy tool like Fiddler/Charles or just use `about:debugging` to inspect the background page and its network requests?

**Comment by bumpySnake** [comment]
simple.py:
```
from http.server import HTTPServer, SimpleHTTPRequestHandler

class Handler(SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header('Set-Cookie', 'mycookie=0; Path=/')
        super().end_headers()

s = HTTPServer(('localhost', 8080), Handler)
s.serve_forever()
```
simple.user.js:
```
// ==UserScript==
// @name DuplicateCookieExample
// @namespace Violentmonkey Scripts
// @match http://localhost:8080/*
// @grant GM_xmlhttpRequest
// ==/UserScript==
GM_xmlhttpRequest({method:'GET', url:'http://localhost:8080'});
```
I go to `about:debugging#/runtime/this-firefox` and click inspect violentmonkey.
xhr message copied from debug console:
```
14:04:48.479 XHRGEThttp://localhost:8080/
[HTTP/1.0 200 OK 1ms]

Request URL:http://localhost:8080/
Request Method:GET
Remote Address:127.0.0.1:8080
Status Code:
200
Version:HTTP/1.0
	
Response Headers (186 B)	
Raw Headers
HTTP/1.0 200 OK
Server: SimpleHTTP/0.6 Python/3.7.2
Date: Thu, 26 Dec 2019 19:04:48 GMT
Content-type: text/html; charset=utf-8
Content-Length: 393
Set-Cookie: mycookie=0; Path=/
Request Headers (275 B)	
Raw Headers
Host: localhost:8080
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:71.0) Gecko/20100101 Firefox/71.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
DNT: 1
Connection: keep-alive
Cookie: mycookie=0, mycookie=0;
```

Sorry to be unclear. Is this helpful?

**Comment by tophf** [comment]
Able to repro, thanks, investigating...

**Comment by tophf** [comment]
So yeah FF merges our `cookie` header with the built-in one using a comma, might be even a bug in FF, but I'm too lazy to investigate it further. I'll just drop the duplicate header.

**Comment by tophf** [comment]
As for the private mode problem, I don't think I understood it, but it could be a separate thing so see if the code from the PR fixes it ([dist.zip](https://github.com/violentmonkey/violentmonkey/files/4003348/dist.zip)), otherwise please open a new issue.

**Comment by gera2ld** [comment]
I did some quick test and it seems that Firefox also sends Cookies natively. I cannot even reproduce #606 .
I will dig further to see if it was a bug of Firefox and has been fixed. If that's the case, I think we can just revert the send-cookies-explicitly part.

**Comment by tophf** [comment]
FF doesn't send cookies natively on sites I tried like youtube.com.

**Comment by gera2ld** [comment]
I think the issue is not about private context, it's because you are visiting `http://localhost:8080`, cookies of which are not blocked by Firefox. In case of `https://www.google.com` there will be no duplicate cookies, and there are warnings:

> Request to access cookie or storage on “https://www.google.com/” was blocked because it came from a tracker and content blocking is enabled.

So I think the real problem is, whether we should inject cookies manually depends on whether Firefox blocks them, when identifying it coming from a tracker. Currently the decision on whether to inject cookies is unreliable, leading to duplicate cookies in some cases.
